packages/mui-joy/src/Drawer/Drawer.tsx

```diff
@@ -0,0 +1,213 @@
+import { useThemeProps, Theme } from '../styles';
+import styled from '../styles/styled';
+import { getDrawerUtilityClass } from './drawerClasses';
+import { DrawerProps, DrawerOwnerState, DrawerTypeMap } from './DrawerProps';
+import useSlot from '../utils/useSlot';
+
+const useUtilityClasses = () => {
+  const slots = {
+    root: ['root'],
+    sheet: ['sheet'],
+  };
+
+  return composeClasses(slots, getDrawerUtilityClass, {});
+};
+
+const DrawerRoot = styled(Modal, {
+  name: 'JoyDrawer',
+  slot: 'Root',
+  overridesResolver: (props, styles) => styles.root,
+  // Propagate all props to the underlaying Modal
+  shouldForwardProp: () => true,
+})<{ ownerState: DrawerOwnerState }>(({ ownerState }) => {
+  return [
+    {
+      transitionProperty: 'visibility',
+      transitionDelay: ownerState.open ? '0s' : '300ms',
+      [`& .${modalClasses.backdrop}`]: {
+        opacity: ownerState.open ? 1 : 0,
+        transition: 'opacity 0.3s ease',
+      },
+    },
+  ];
+});
+
+const DrawerSheet = styled(Sheet, {
+  name: 'JoyDrawer',
+  slot: 'Sheet',
+  overridesResolver: (props, styles) => styles.root,
+})<{ ownerState: DrawerOwnerState }>(({ ownerState }) => {
+  return {
+    boxSizing: 'border-box',
+    position: 'fixed',
+    overflow: 'auto',
+    ...(ownerState.anchor === 'left' && {
+      left: 0,
+      transform: ownerState.open ? 'translateX(0)' : 'translateX(-100%)',
+    }),
+    ...(ownerState.anchor === 'right' && {
+      right: 0,
+      transform: ownerState.open ? 'translateX(0)' : 'translateX(100%)',
+    }),
+    ...(ownerState.anchor === 'top' && {
+      top: 0,
+      transform: ownerState.open ? 'translateY(0)' : 'translateY(-100%)',
+    }),
+    ...(ownerState.anchor === 'bottom' && {
+      bottom: 0,
+      transform: ownerState.open ? 'translateY(0)' : 'translateY(100%)',
+    }),
+    height: ownerState.anchor!.match(/(left|right)/) ? '100%' : 'clamp(256px, 30vw, 378px)',
+    width: ownerState.anchor!.match(/(top|bottom)/) ? '100vw' : 'clamp(256px, 30vw, 378px)',
+    transition: 'transform 0.3s ease',
+  };
+});
+
+const oppositeDirection = {
+  left: 'right',
+  right: 'left',
+  top: 'down',
+  bottom: 'up',
+};
+
+export function isHorizontal(anchor: DrawerProps['anchor']) {
+  return ['left', 'right'].indexOf(anchor!) !== -1;
+}
+
+export function getAnchor(theme: Theme, anchor: DrawerProps['anchor']) {
+  // @ts-ignore TODO: How is the direction set in Joy UI?
+  return theme.direction === 'rtl' && isHorizontal(anchor) ? oppositeDirection[anchor] : anchor;
+}
+
+/**
+ * The navigation drawers (or "sidebars") provide ergonomic access to destinations in a site or app functionality such as switching accounts.
+ *
+ * Demos:
+ *
+ * - [Drawer](https://mui.com/joy-ui/react-drawer/)
+ *
+ * API:
+ *
+ * - [Drawer API](https://mui.com/joy-ui/api/drawer/)
+ * - inherits [Modal API](https://mui.com/joy-ui/api/modal/)
+ */
+const Drawer = React.forwardRef(function Drawer(inProps, ref) {
```

Comment:
It's less than 60 lines of code with component composition, in comparison with the modal for e.g. that has around 200 lines of code. This is only to compare the complexity